src/parser.cpp

```diff
@@ -174,61 +168,61 @@ namespace ada::parser {
 
           if(!contains_ampersand) {
             state = ada::state::HOST;
-            goto goto_host;
-          }
-          bool at_sign_seen{false};
-          bool password_token_seen{false};
-          do {
-            std::string_view view = helpers::substring(url_data, input_position);
-            size_t location = url.is_special() ? view.find_first_of("@/?\\") : view.find_first_of("@/?");
-            std::string_view authority_view(view.data(), (location != std::string_view::npos) ? location : view.size());
-            input_position = (location == std::string_view::npos) ? input_size : input_position + location;
-            // If c is U+0040 (@), then:
-            if ((input_position != input_size) && (url_data[input_position] == '@')) {
-              // If atSignSeen is true, then prepend "%40" to buffer.
-              if (at_sign_seen) {
-                if (password_token_seen) {
-                  url.password += "%40";
-                } else {
-                  url.username += "%40";
+          } else {
+            bool at_sign_seen{false};
+            bool password_token_seen{false};
+            do {
+              std::string_view view = helpers::substring(url_data, input_position);
+              size_t location = url.is_special() ? view.find_first_of("@/?\\") : view.find_first_of("@/?");
+              std::string_view authority_view(view.data(), (location != std::string_view::npos) ? location : view.size());
+              input_position = (location == std::string_view::npos) ? input_size : input_position + location;
```

Comment:
Here it is maybe unwise to increment input_position unconditionally. It might be simpler to only increment it when you won't have to later decrement it.